R/paginate_listing.R

```diff
@@ -49,20 +49,27 @@ paginate_listing <- function(lsting,
                              tf_wrap = !is.null(max_width),
                              rep_cols = NULL,
                              max_width = NULL,
+                             fontspec = font_spec(font_family, font_size, lineheight),
                              verbose = FALSE,
                              print_pages = TRUE) {
-  checkmate::assert_multi_class(lsting, c("listing_df", "list"))
-  checkmate::assert_numeric(colwidths, lower = 0, len = length(listing_dispcols(lsting)), null.ok = TRUE)
```

Comment:
We support paginating lists so these checks need to be preserved

Reply:
I'm not sure "paginating lists" is well defined. If the goal is to support `split_into_pages_by_var`, that could set something that a `do_forced_pag` method for listings does, which would ensure things like the fontspec, the colwidths, the col_gap, etc are all the same at pagination time, which carrying them around in a list doesn't.

The issue is that technically speaking, I can make 3 listings and put them into a list and pass it to `paginate_listing` but they're not really pages of the same list at that point, and thus, e.g., probably shouldn't ahve the same page numbering, etc. 

My suggestion above can be implemented without changing the user facing API if `split_into_pages_by_var` is already in use, which I suspect it is

Reply:
It would look something like (untested but should either work or be very close to it). Note this also a) uses page title instead of subtitles to carry around the page titles, so it won't clobber a subtitle that is already on the parent listing, and b) this support splitting by more than one variable, if desired.

```
split_into_pages_by_var2 <- function(lsting, cols, page_prefix = cols ) {
   stopifnot(length(cols) == length(page_prefix),
                    all(cols %in% listing_dispcols(lsting)))
  ldf_page_by_vars(lsting) <- cols
  ldf_page_prefixes(lsting) <- page_prefix
  lsting
}

setMethod(do_forced_paginate, "listing_df",
          function(obj) {
    paglst <- list(obj)
    pbvars <- ldf_page_by_vars(obj)
    prefixes <- ldf_page_prefixes(obj)
    for(i in seq_along(pbvars)) {
      splvar <- pbvars[i]
      prefix <-   prefixes[i]
      paglst <- unlist(
          lapply(paglst,
                 function(pg) {
              splpgs <- split(pg, pg[[splvar]])
              lapply(splgpgs,
                     function(pg) {
                  page_titles(pg) <- c(page_titles(pg),
                                       paste(preix, ":", pg[[splvar]][1]))
                  pg
              }),
          }),
          recursive = FALSE)
    }
    paglst
})

ldf_page_by_vars <- function(lsting) {
    stopifnot(is(lsting, "listing_df"))
    attr(lsting, "page_by_vars")
}

`ldf_page_by_vars<-` <- function(lsting, value) {
    stopifnot(is(lsting, "listing_df"),
              all(value %in% listing_dispcols(lsting)))
    attr(lsting, "page_by_vars") <- value
    lsting
}

ldf_page_prefixes <- function(lsting) {
    stopifnot(is(lsting, "listing_df"))
    attr(lsting, "page_by_vars")
}


`ldf_page_prefixes` <- function(lsting, value) {
    stopifnot(is(lsting, "listing_df"),
              is.null(value) || (is.character(value) && length(value) == length(ldf_page_by_vars)))
    attr(lsting, "page_prefixes") <- value
    lsting
}

```

Reply:
Yes, the feature was requested some time ago, and we implemented it as a parameter in `as_listing` called `split_into_pages_by_var=`. Docs are available `?split_into_pages_by_var`. The pagination of lists is supported in paginate_to_mpfs at the moment, taking into account page numbering and if we want to print mixed listings/tables.

I suggest moving this discussion to an issue/PR where this feature can be reconsidered. Meanwhile, I would revert these changes as they affect checks made for that feature